tests/framework/extensions/provisioning/creates.go

```diff
@@ -557,3 +600,75 @@ func setLogrusFormatter() {
+	}
+	if nodePool.SpecifyCustomPrivateIP {
+		command += fmt.Sprintf(" --internal-address %s", privateIP)
+	}
+	if nodePool.CustomNodeNameSuffix != "" {
+		command += fmt.Sprintf(" --node-name %s", namegen.AppendRandomString(nodePool.CustomNodeNameSuffix))
+	}
+	for labelKey, labelValue := range nodePool.NodeLabels {
+		command += fmt.Sprintf(" --label %s=%s", labelKey, labelValue)
+	}
+	for _, taint := range nodePool.NodeTaints {
+		command += fmt.Sprintf(" --taints %s=%s:%s", taint.Key, taint.Value, taint.Effect)
+	}
+	return command
+}
+
+// createRegistrationCommand is a helper for rke2/k3s custom clusters to create the registration command with advanced options configured per node
+func createRegistrationCommand(command, publicIP, privateIP string, machinePool provisioninginput.MachinePools) string {
+	if machinePool.SpecifyCustomPublicIP {
+		command += fmt.Sprintf(" --address %s", publicIP)
+	}
+	if machinePool.SpecifyCustomPrivateIP {
+		command += fmt.Sprintf(" --internal-address %s", privateIP)
+	}
+	if machinePool.CustomNodeNameSuffix != "" {
+		command += fmt.Sprintf(" --node-name %s", namegen.AppendRandomString(machinePool.CustomNodeNameSuffix))
+	}
+	for labelKey, labelValue := range machinePool.NodeLabels {
+		command += fmt.Sprintf(" --label %s=%s", labelKey, labelValue)
+	}
+	for _, taint := range machinePool.NodeTaints {
+		command += fmt.Sprintf(" --taints %s=%s:%s", taint.Key, taint.Value, taint.Effect)
+	}
+	return command
+}
+
+// createWindowsRegistrationCommand is a helper for rke2 windows custom clusters to create the registration command with advanced options configured per node
+func createWindowsRegistrationCommand(command, publicIP, privateIP string, machinePool provisioninginput.MachinePools) string {
+	if machinePool.SpecifyCustomPublicIP {
+		command += fmt.Sprintf(" -Address '%s'", publicIP)
+	}
+	if machinePool.SpecifyCustomPrivateIP {
+		command += fmt.Sprintf(" -InternalAddress '%s'", privateIP)
+	}
+	if machinePool.CustomNodeNameSuffix != "" {
+		command += fmt.Sprintf(" -NodeName '%s'", namegen.AppendRandomString(machinePool.CustomNodeNameSuffix))
+	}
+	// powershell requires only 1 flag per command, so we need to append the custom labels and taints together which is different from linux
+	if len(machinePool.NodeLabels) > 0 {
+		// there is an existing label for all windows nodes, so we need to insert the custom labels after the existing label
+		labelIndex := strings.Index(command, " -Label '") + len(" -Label '")
+		customLabels := ""
+		for labelKey, labelValue := range machinePool.NodeLabels {
+			customLabels += fmt.Sprintf("%s=%s,", labelKey, labelValue)
+		}
+		command = command[:labelIndex] + customLabels + command[labelIndex:]
+	}
+	if len(machinePool.NodeTaints) > 0 {
+		customTaints := " -Taint '"
+		for _, taint := range machinePool.NodeTaints {
+			customTaints += fmt.Sprintf("%s=%s:%s,", taint.Key, taint.Value, taint.Effect)
+		}
+		customTaints += "'"
```

Comment:
Maybe something like this - can be like below or with any other approach to prevent single quotes:

```suggestion
		var customTaints string
		for _, taint := range machinePool.NodeTaints {
			customTaints += fmt.Sprintf("%s=%s:%s,", taint.Key, taint.Value, taint.Effect)
		}
		wrappedTaints := fmt.Sprintf(" -Taint '%s'", customTaints)
```

Probably a nit, this way we can prevent any by-accident `'` removals from the `customTaints`

Reply:
good point.

Reply:
this has been updated - much prettier thanks for the recommendation